src/scaffoldfitter/fitterstepalign.py

```diff
@@ -275,32 +279,71 @@ def _optimiseAlignment(self, pointMap):
             nodetemplate.defineField(modelCoordinates)
             nodetemplate.defineField(dataCoordinates)
             fieldcache = fieldmodule.createFieldcache()
+
+            modelsum = [0.0, 0.0, 0.0]
+            datasum = [0.0, 0.0, 0.0]
             for name, positions in pointMap.items():
                 modelx = positions[0]
                 datax = positions[1]
+                modelsum = [modelsum[c] + modelx[c] for c in range(3)]
+                datasum = [datasum[c] + datax[c] for c in range(3)]
                 node = nodes.createNode(-1, nodetemplate)
                 fieldcache.setNode(node)
                 result1 = modelCoordinates.assignReal(fieldcache, positions[0])
                 result2 = dataCoordinates.assignReal(fieldcache, positions[1])
                 assert (result1 == RESULT_OK) and (result2 == RESULT_OK), "Align:  Failed to set up data for alignment to markers optimisation"
+
+            groupCount = len(pointMap)
+            modelCM = [c / groupCount for c in modelsum]
+            dataCM = [c / groupCount for c in datasum]
+            translationOffset = [dataCM[c] - modelCM[c] for c in range(3)]
+
             del fieldcache
-            modelCoordinatesTransformed, rotation, scale, translation = createFieldsTransformations(modelCoordinates, translation_scale_factor=dataScale)
-            # create objective = sum of squares of vector from modelCoordinatesTransformed to dataCoordinates
-            markerDiff = fieldmodule.createFieldSubtract(dataCoordinates, modelCoordinatesTransformed)
-            scaledMarkerDiff = markerDiff*fieldmodule.createFieldConstant([ 1.0/dataScale ]*3)
-            objective = fieldmodule.createFieldNodesetSumSquares(scaledMarkerDiff, nodes)
-            #objective = fieldmodule.createFieldNodesetSum(fieldmodule.createFieldMagnitude(scaledMarkerDiff), nodes)
-            assert objective.isValid(), "Align:  Failed to set up objective function for alignment to markers optimisation"
 
-        # future: pre-fit to avoid gimbal lock
+            # Pre-align to avoid gimbal lock
+            translationScaleFactor = 1.0
+            for x in range(2):
+                for y in (range(4) if x == 0 else (0, 2)):
+                    for z in range(4):
+                        azimuth = 0.5 * math.pi * z
+                        elevation = 0.5 * math.pi * y
+                        roll = 0.5 * math.pi * x
+                        rotationAngles = [azimuth, elevation, roll]
+
+                        modelCoordinatesTransformed, rotation, scale, translation = createFieldsTransformations(
+                            modelCoordinates, rotationAngles, scaleFactor, translationOffset, translationScaleFactor)
```

Comment:
Translations are a pain and this is the crunch point: translationOffset is only valid at zero rotation angles.
Here we apply it as translate(scale(rotate(modelcoordinates))), so while it's true the translation is applied last, the problem is that the model centre of mass is subject to the rotation and scale. Normally we overcome this by subtracting modelCM first, doing rotation and scale, then adding the dataCM last. However, the optimisation doesn't have a unique solution if it involves two translations & I'm not wanting to complicate the UI by having two either.
There's no need to define the fields repeatedly in this loop; you can define them all outside, and all you need to do is assign rotation with ``rotation.assignReal(fieldcache, rotationAngles)`` and similarly for translation.
opencmiss.utils has a method to get the transformation matrix in 3x3 python list of lists; can use this to add a correction to the translation:
```
from opencmiss.utils.maths.vectorops import add, euler_to_rotation_matrix, matrix_mult, mult, sub

rotationMatrix = euler_to_rotation_matrix(rotationAngles)
translationFix = mult(sub(matrix_mult(rotationMatrix, modelCM), modelCM), scaleFactor)
translation.assignReal(fieldcache, add(translationOffset, translationFix))
```
Please check my maths! It's also worth confirming that different rotations are being accepted as the best starting point; I suspect the current code would make the zero rotation match most often.